hashintel/hash
# H-206: Allow for insertion of property/link values into pages

## 🌟 What is the purpose of this PR?




This PR adds an improved version of the mention suggestor, which allows for inserting entities, entity property values, or entity outgoing link values into a page.


## 🔗 Related links





- H-206

## 🔍 What does this change?





## Pre-Merge Checklist 🚀

### 🚢 Has this modified a publishable library?




This PR:


### 📜 Does this require a change to the docs?




The changes in this PR:


### 🕸️ Does this require a change to the Turbo Graph?




The changes in this PR:


## ⚠️ Known issues



- There are numerous issues with the current implementation of the search input which require addressing (have created H-862 for this):
  - the cursor in the search input is not styled correctly
  - there is missing padding between the `@` and the search text in the search input
  - there is a react error that is thrown occasionally when switching between pages, which I think was introduced as part of the implementation of the search input, which creates nodes in the DOM that aren't controlled by react. It is possible this may go away if we make use of `Decoration.widget` instead of `Decoration.inline` to style the input, as this allows us to use custom react elements to add the styling/functionality of the search input. 

## 🐾 Next steps



- H-862
- H-930
- H-929
- H-928

## 🛡 What tests cover this?



None

## ❓ How to test this?



Because this PR touches `libs/@local/hash-isomorphic-utils/src/prosemirror.ts`, which is used in the BE, it's my understanding that it needs to be tested locally rather than in the deployment.

1. Go to a page
2. Type `@`, which should display a search input and the mention suggester dropdown.
3. To select an entity, click on one or press enter. To select a property or an outgoing link, use the keyboard arrows or click the right chevron to open the entity's sub-menu and select the corresponding value there.
4. Observe the link/property/entity being inserted into the document. When clicking on property/link values, a popup should appear indicating the corresponding link/property type, and source entity.

## 📹 Demo




https://github.com/hashintel/hash/assets/42802102/bfdc7a2e-9902-4edc-957a-727dfdf9a1f1
